File: baekjoon_all/13000+/boj_13335.java
// Solve 2024-01-27

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class boj_13335 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int n = Integer.parseInt(st.nextToken());
        int w = Integer.parseInt(st.nextToken());
        int l = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine(), " ");
        Queue<Integer> trucksIn = new LinkedList<>();

        for (int i = 0; i < n; i++) {
            trucksIn.add(Integer.parseInt(st.nextToken()));
        }

        int curTime = 0;
        int curWeightSum = 0;
        Queue<int[]> trucksOut = new LinkedList<>();

        while (true) {
            curTime++;

            if (!trucksOut.isEmpty() && trucksOut.peek()[1] == curTime) {
                curWeightSum -= trucksOut.poll()[0];
            }

            if (!trucksIn.isEmpty() && curWeightSum + trucksIn.peek() <= l) {
                int truck = trucksIn.poll();
                curWeightSum += truck;
                trucksOut.add(new int[] { truck, curTime + w });
            }

            if (trucksIn.isEmpty() && trucksOut.isEmpty()) break;
        }

        System.out.println(curTime);
    }

}

File: baekjoon_all/13000+/boj_13335.cpp
// Solve 2024-01-28

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;

int main() {
    FASTIO;

    int n, w, l;
    cin >> n >> w >> l;

    queue<int> trucks_in;

    for (int i = 0; i < n; i++) {
        int truck;
        cin >> truck;

        trucks_in.push(truck);
    }

    queue<pair<int, int>> trucks_out;
    int cur_time = 0;
    int cur_weight_sum = 0;

    while (true) {
        cur_time++;

        if (!trucks_out.empty() && trucks_out.front().second == cur_time) {
            cur_weight_sum -= trucks_out.front().first;
            trucks_out.pop();
        }

        if (!trucks_in.empty() && cur_weight_sum + trucks_in.front() <= l) {
            int truck = trucks_in.front();
            trucks_in.pop();
            cur_weight_sum += truck;
            trucks_out.emplace(truck, cur_time + w);
        }

        if (trucks_in.empty() && trucks_out.empty()) break;
    }

    cout << cur_time << '\n';

    return 0;
}
